service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java

```diff
@@ -177,9 +177,22 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
     ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
 
     try {
+      VirtualMachine resolvedVm =
+              computeManager
+                      .virtualMachines()
+                      .getByResourceGroup(
+                              azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
       computeManager
-          .virtualMachines()
-          .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+              .virtualMachines()
+              .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
```

Comment:
I am not sure I understand the dependencies, but I think this undo code is not idempotent. If the pod crashed (or was stopped by K8s) at this point, then a rerun of this undo would not be able to locate the nics or the disks. The VM would be deleted so the retrieval at line 180 would not find it.

The fix would be to find the objects by name, the same way you find them in the `do()` code.

Reply:
good point..I can fix it for nic, but OS disk doesn't have a known name, so Idk if we can fix this completely

Reply:
Probably my misunderstanding about the disk. I see at line 85, a disk is retrieved by name. Is that separate from the OS disk? Or is that the disk you are supposed to delete?

If the OS disk is automatically created by Azure, but not automatically deleted by Azure, I don't see a way to fix the idempotency.

Reply:
>Probably my misunderstanding about the disk. I see at line 85, a disk is retrieved by name. Is that separate from the OS disk? Or is that the disk you are supposed to delete?

yea OS disk and the user disk are 2 disks. We define user disk but not OS disk.

>If the OS disk is automatically created by Azure, but not automatically deleted by Azure, I don't see a way to fix the idempotency.
yes.

Reply:
OK, then would you document the problem case (orphan NIC and disk in a small failure window) in the undo code for our future selves? I'll approve the PR now.